### ISSUE TITLE
[Question] Switch picker without leaving fzf-window

### ISSUE THREAD

**Issue by Bekaboo** [ibhagwan/fzf-lua#974]
### Info

Not related

### Description 

First of all thanks for making such a great plugin. fzf-lua is fast, responsive, and feature-rich, I cannot use neovim without it.

I sometimes find myself using the wrong picker only after I have typed several words in the search prompt, e.g. I was meant to live grep patterns in the project but I opened fzf files picker. In this case I want to switch to the live grep picker without leaving fzf window and perserve the previous `cwd` and the contents in the search prompt, but I haven't found a good way to achieve that.

Currently, I can map `<C-_>` in the `on_create` callback to `fzf.builtin` to switch picker without leaving fzf window, like this:

```lua
local fzf = require('fzf-lua')

fzf.setup({
  winopts = {
    on_create = function()
      vim.keymap.set('t', '<C-_>', fzf.builtin, { buffer = true })
    end,
  },
})
```

But this method has two drawbacks:

1. It does not preserve the contents in the search prompt
2. It does not keep previous search cwd

Could you please give me some hints or guidance on how to achieve this? Thanks!

**Comment by ibhagwan** [comment]
> First of all thanks for making such a great plugin. fzf-lua is fast, responsive, and feature-rich, I cannot use neovim without it.

Ty for your kind words @Bekaboo!
 
Your request is possible but it would require a bit more involved code as you’ll need to pass the new options to the new provider.
 
> It does not preserve the contents in the search prompt

You can access the last query with `require('fzf-lua').get_info().last_query`.

> It does not keep previous search cwd

While a bit more hacky, the last `cwd` (and all other call options) can be found in the last resume options `require('fzf-lua').config.__resume_opts.cwd`

**Comment by Bekaboo** [comment]
Hi thanks for your reply, after searching in the source code I find another function `require('fzf-lua').get_last_query()` is it the suggested way of getting the last query than `require('fzf-lua').get_info().last_query()`?

I tried them both but none of them seems to work, here is my config:

```lua
local fzf = require('fzf-lua')

fzf.setup({
  winopts = {
    on_create = function()
      vim.keymap.set('t', '<C-_>', function()
        -- fzf.builtin()
        -- vim.api.nvim_feedkeys('', 'n', true)
        local info_last_query = fzf.get_info().last_query
        local last_query = fzf.get_last_query()
        local resume_opts = fzf.config.__resume_opts
        print('info_last_query: ' .. vim.inspect(info_last_query))
        print('last_query: ' .. vim.inspect(last_query))
        print('resume_opts: ' .. vim.inspect(resume_opts))
      end, { buffer = true })
    end,
  },
})
```

All three values (`info_last_query`, `last_query`, and `resume_opts`) are `nil` after I press `<C-_>` in fzf window. I've tried to call `fzf.builtin()` first and close the fzf window first but that does not help. What am I doing wrong here?

**Comment by ibhagwan** [comment]
You’re correct `get_last_query` is better, I think info doesn’t even have the last query, my bad :)

The issue we have here is that you’re opening the `builtin` interface which also overrides the last query, there is also `config.__resume_map` which has resume values per provider but that means you’ll need to know which was the last interface used, I’ll think of something and do some testing and let you know how we can achieve what you’re after.

**Comment by ibhagwan** [comment]
Another issue I see here is that you’re using a mapping in `winopts.on_create`, this abruptly closes the interface and opens another instead which also prevents the last query from being stored, this is better done via an `action` mapping.

**Comment by ibhagwan** [comment]
The below adds the functionality you're after for all files providers, if you wish to do that for buffers as well you'd need to also configure `actions.buffers`:
```lua
local fzf = require('fzf-lua')
local actions = fzf.actions

fzf.setup({
  actions = {
    files = {
      ["default"] = actions.file_edit_or_qf,
      ["ctrl-s"]  = actions.file_split,
      ["ctrl-v"]  = actions.file_vsplit,
      ["ctrl-t"]  = actions.file_tabedit,
      ["alt-q"]   = actions.file_sel_to_qf,
      ["alt-l"]   = actions.file_sel_to_ll,
      ["ctrl-_"]  = {
        function()
          local opts = {
            query = fzf.config.__resume_data.last_query,
            cwd = fzf.config.__resume_data.opts.cwd,
          }
          fzf.builtin({
            actions = {
              ["default"] = function(selected)
                fzf[selected[1]](opts)
              end,
            }
          })
        end
      },
    },
  },
})

```

**Comment by Bekaboo** [comment]
Thanks, it works incredibly well!

**Comment by hexh250786313** [comment]
I encountered a similar problem, but with a slight difference. I attempted to create a finder that displays the MRU list when the query is empty, and shows the results from `fd` when the query is not empty.

I wasn't sure how to handle this 'switcher' with fzf-lua, so I ended up implementing it using the vanilla junegunn/fzf, using `--bind=change:reload(dynamic_fzf_source.sh)`

If possible, could anyone offer me some suggestions?

**Comment by ibhagwan** [comment]
@hexh250786313, you are tryin to mix and match a “live” reload mechanism with fuzzy matching, while possible it’s a bit tricky, if we ignore for a second displaying the MRU on backspace you can change back to fuzzy with `enable-search`:
```sh
--bind "change:reload(fd …)+unbind(change)+change-prompt(fd> )+enable-search"
```

It might be possible to create go back to the MRU on backspace with a condition that tests the query with `{q}` but that means you can’t unbind `change`. This is just atop of my head, I might be able to come up with something better if I think about this longer :)

**Comment by dan-cooke** [comment]
@hexh250786313

I have been searching far and wide for this behavior, by any chance could you share your dotfiles for this?

**Comment by dan-cooke** [comment]
@ibhagwan is it possible to achieve this kind of thing within fzf-lua?

**Comment by ibhagwan** [comment]
> @ibhagwan is it possible to achieve this kind of thing within fzf-lua?

Not really sure what you’re asking, this is an old issue, by “this kind of thing” do you mean the OP/title or something else? Whatever the question may be it probably is possible depending on how much effort you’re willing to invest.

**Comment by dan-cooke** [comment]
@ibhagwan apologies I mean the MRU/files switch that @hexh250786313 was asking about

**Comment by ibhagwan** [comment]
@dan-cooke, my answer https://github.com/ibhagwan/fzf-lua/issues/974#issuecomment-1873990328 still stands, I can also think of another solution using the Linux headless wrapper used by fzf-lua, take a look at `rg_glob_fn` from #373, but then again it won’t be fuzzy search since it’s for live_grep.

I can think of many different approaches, all requires to make use of fzf’s `change:…` event, if you’re serious about it you’ll have to do some reading of the fzf manual and issues to understand how to use it and what’s possible, anything that can be done with vanilla fzf can be done with fzf-lua (with richer functionality), once you have basic code that does what you want I can help you refine it.

**Comment by dan-cooke** [comment]
@ibhagwan thanks so much for your helpful replies! I am quite new to Lua/Neovim plugins so I wasn't quite sure how to call the underlying fzf from neovim. 

I had a look at @hexh250786313 's dotfiles and I think I'm on the right track.

I will post my solution in here when I'm done!

**Comment by dan-cooke** [comment]
@ibhagwan so I have this
```.lua
vim.keymap.set('n', '<C-p>', function()
	fzf.oldfiles({
		prompt = 'Recently Used > ',
		path_shorten = true,
		file_name_only = true,
		cwd_only = true,
		keymap = {
			-- These override the default tables completely
			-- no need to set to `false` to disable a bind
			-- delete or modify is sufficient
			builtin = {
				-- neovim `:tmap` mappings for the fzf win
				["<F1>"]     = "toggle-help",
				["<F2>"]     = "toggle-fullscreen",
				-- Only valid with the 'builtin' previewer
				["<F3>"]     = "toggle-preview-wrap",
				["<F4>"]     = "toggle-preview",
				-- Rotate preview clockwise/counter-clockwise
				["<F5>"]     = "toggle-preview-ccw",
				["<F6>"]     = "toggle-preview-cw",
				["<S-down>"] = "preview-page-down",
				["<S-up>"]   = "preview-page-up",
				["<S-left>"] = "preview-page-reset",
				["<C-p>"]    = "down",
				["<C-S-p>"]  = "up",
			},
			fzf = {
				-- fzf '--bind=' options
				["ctrl-z"]     = "abort",
				["ctrl-u"]     = "unix-line-discard",
				["ctrl-f"]     = "half-page-down",
				["ctrl-b"]     = "half-page-up",
				["ctrl-a"]     = "beginning-of-line",
				["ctrl-e"]     = "end-of-line",
				["alt-a"]      = "toggle-all",
				["change"]     = "reload(fd …)+unbind(change)+change-prompt(fd> )+enable-search",
				-- Only valid with fzf previewers (bat/cat/git/etc)
				["f3"]         = "toggle-preview-wrap",
				["f4"]         = "toggle-preview",
				["shift-down"] = "preview-page-down",
				["shift-up"]   = "preview-page-up",
				["ctrl-p"]     = "down",
				["ctrl-n"]     = "up"
			},
		},


	})
end)
```

But there is a problem, the fd search results are always empty
![image](https://github.com/ibhagwan/fzf-lua/assets/22816887/20fcc804-ccd8-434f-9e13-92607f0f3b16)


Any ideas?

**Comment by ibhagwan** [comment]
> But when I switch from oldfiles to fd - the old files is not updated with the entry I select, is there something I can do about that?

That’s because the shada file isn’t written to disk when you open a file, it is online added to the shada when you exit neovim.

>` include_current_session = true`

 This is a good solution, another way would be to an autocmd that updates the shada file on `BufReadPost`.

**Comment by mangelozzi** [comment]
This is great info, thanks @ibhagwan !

For any future people like me when you start a new sessions and you try find a buffer, but you havent opened it yet, you can switch to the file picker with `CTRL+k` like this (also includes switching back to buffers again with `<Ctrl+k>`):

```lua
        actions = {
            buffers = { -- When on :buffers, pressing <Ctrl-k> will resume with files
                ["ctrl-k"]  = {
                    function()
                        local opts = {
                            query = fzf.config.__resume_data.last_query,
                            cwd = fzf.config.__resume_data.opts.cwd,
                        }
                        fzf.files(opts)
                    end
                },
            },
            files = { -- When on :files, pressing <Ctrl-k> will resume with buffers
                ["ctrl-k"]  = {
                    function()
                        local opts = {
                            query = fzf.config.__resume_data.last_query,
                            cwd = fzf.config.__resume_data.opts.cwd,
                        }
                        fzf.buffers(opts)
                    end
                },
            },
        },
```

**Comment by ibhagwan** [comment]
Ty @mangelozzi, btw, I’ve made the last query more accessible from actions:
```lua

["ctrl-k"]  = function(sel, opts)
  print('q:', opts.last_query)
end
```

**Comment by mangelozzi** [comment]
> Ty @mangelozzi, btw, I’ve made the last query more accessible from actions:

Thanks! I made it more compact with your suggestion:

```lua
        actions = {
            buffers = { -- When on :buffers, pressing <Ctrl-Space> will resume with files
                ["Ctrl-Space"] = function(_, opts) fzf.files({ query = opts.last_query }) end, -- _ = sel
            },
            files = { -- When on :files, pressing <Ctrl-Space> will resume with buffers
                ["Ctrl-Space"] = function(_, opts) fzf.buffers({ query = opts.last_query }) end, -- _ = sel
            }
        },
```

**Comment by ibhagwan** [comment]
Looks great @mangelozzi!
